src/survey_assist_utils/evaluation/coder_alignment.py

```diff
@@ -0,0 +1,550 @@
+        """Cleans the DataFrame by handling data types and missing values."""
+        # Convert all label columns to string type first
+        label_cols = self.model_label_cols + self.clerical_label_cols
+        df[label_cols] = df[label_cols].astype(str)
+
+        # Replace all "impostor NaNs" in label columns with a true NaN
+        df[label_cols] = df[label_cols].replace(self._MISSING_VALUE_FORMATS, np.nan)
+
+        # Add leading zeros, whilst maintaining nans
+        for col in label_cols:
+            # Apply zfill to non na/nan
+            df[col] = df[col].apply(self._safe_zfill)
+
+        return df
+
+    def _melt_and_clean(self, value_vars: list[str], value_name: str) -> pd.DataFrame:
+        """Helper to reshape data from wide to long and drop any remaining NaNs."""
+        melted_df = self.df.melt(
+            id_vars=[self.id_col], value_vars=value_vars, value_name=value_name
+        )
+        # Now we only need to drop true NaNs, as cleaning was done in __init__
+        return melted_df.dropna(subset=[value_name])
+
+    def _add_derived_columns(self):
+        """Adds computed columns for full and partial matches (vectorised)."""
+        model_melted = self._melt_and_clean(self.model_label_cols, "model_label")
+        clerical_melted = self._melt_and_clean(
+            self.clerical_label_cols, "clerical_label"
+        )
+
+        full_matches = pd.merge(
+            model_melted,
+            clerical_melted,
+            left_on=[self.id_col, "model_label"],
+            right_on=[self.id_col, "clerical_label"],
+        )
+        full_match_ids = full_matches[self.id_col].unique()
+
+        model_melted["model_label_2_digit"] = model_melted["model_label"].str[:2]
+        clerical_melted["clerical_label_2_digit"] = clerical_melted[
+            "clerical_label"
+        ].str[:2]
+
+        partial_matches = pd.merge(
+            model_melted,
+            clerical_melted,
+            left_on=[self.id_col, "model_label_2_digit"],
+            right_on=[self.id_col, "clerical_label_2_digit"],
+        )
+        partial_match_ids = partial_matches[self.id_col].unique()
+
+        self.df["is_correct"] = self.df[self.id_col].isin(full_match_ids)
+        self.df["is_correct_2_digit"] = self.df[self.id_col].isin(partial_match_ids)
+
+        for col in self.model_score_cols:
+            self.df[col] = pd.to_numeric(self.df[col], errors="coerce")
+        self.df["max_score"] = self.df[self.model_score_cols].max(axis=1)
+
+    def get_jaccard_similarity(self) -> float:
+        """Calculates the average Jaccard Similarity Index across all rows."""
+
+        def calculate_jaccard_for_row(row):
+            # No need to clean here, as self.df is already cleaned
+            model_set = set(row[self.model_label_cols].dropna())
+            clerical_set = set(row[self.clerical_label_cols].dropna())
+
+            if not model_set and not clerical_set:
```

Comment:
Not sure I'm understanding the logic for this condition. What scenario is it addressing?

Reply:
In any case, if both sets are empty, you would say that there's nothing in common, e.g. Jaccard Similarity should be 0.